src/requirements.ts

```diff
@@ -66,33 +59,54 @@ function checkJavaRuntime(): Promise<string> {
         }
         invalidJavaHome(reject, msg);
       }
-      return resolve(javaHome);
+      resolve(javaHome);
     }
-    //No settings, let"s try to detect as last resort.
+    // No settings, check if we have a managed one
+    javaHome = context.globalState.get(JAVA_MANAGED_HOME_KEY, null);
+    if (javaHome) {
+      resolve(javaHome);
+    }
+
+    // No settings and no existing managed one, let's try to detect
     findJavaHome((err, home) => {
-      if (err) {
-        openJREDownload(
-          reject,
-          `Java runtime could not be located. Install it and set its location using "${JAVA_HOME_CONFIG}" variable in VS Code settings.`
-        );
+      if (err || !home) {
+        // No Java detected, last resort is to ask for permission to download and manage our own
+        suggestManagedJre(reject);
       } else {
         resolve(home);
       }
     });
   });
 }
 
+function tryExplicitConfiguration() {
+  let source: string;
+  let javaHome: string = readJavaConfig();
+  if (javaHome) {
+    source = `'${JAVA_HOME_CONFIG}' variable defined in VS Code settings`;
+  } else {
+    javaHome = process.env['JDK_HOME'];
+    if (javaHome) {
+      source = 'JDK_HOME environment variable';
+    } else {
+      javaHome = process.env['JAVA_HOME'];
+      source = 'JAVA_HOME environment variable';
+    }
+  }
+  return { source, javaHome: javaHome ? javaHome.trim() : null };
+}
+
 function readJavaConfig(): string {
-  const config = workspace.getConfiguration();
-  return config.get<string>(JAVA_HOME_CONFIG, null);
+  return vscode.workspace.getConfiguration().get<string>(JAVA_HOME_CONFIG, null);
 }
 
 function checkJavaVersion(javaHome: string): Promise<number> {
   return new Promise((resolve, reject) => {
-    cp.execFile(javaHome + '/bin/java', ['-version'], {}, (error, stdout, stderr) => {
+    cp.execFile(`${javaHome}/bin/java`, ['-version'], {}, (error, stdout, stderr) => {
```

Comment:
You used `path.join` somewhere else, could be nice to have it here to (I don't know what happens behind `execFile`).